libs/x-content/src/main/java/org/elasticsearch/common/xcontent/MediaTypeParser.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.xcontent;
+
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+
+public class MediaTypeParser<T extends MediaType> {
+    private Map<String, T> formatToMediaType = new HashMap<>();
+    private Map<String, T> typeSubtypeToMediaType = new HashMap<>();
+
+    public MediaTypeParser(T[] acceptedMediaTypes) {
+        for (T mediaType : acceptedMediaTypes) {
+            typeSubtypeToMediaType.put(mediaType.typeSubtype(), mediaType);
+            formatToMediaType.put(mediaType.format(), mediaType);
+        }
+    }
+
+    public T fromMediaType(String mediaType) {
+        ParsedMediaType parsedMediaType = parseMediaType(mediaType);
+        return parsedMediaType != null ? parsedMediaType.getMediaType() : null;
+    }
+
+    public T fromFormat(String format) {
+        if(format == null) {
+            return null;
+        }
+        return formatToMediaType.get(format.toLowerCase(Locale.ROOT));
+    }
+
+    public MediaTypeParser<T> withAdditionalMediaType(String typeSubtype, T xContentType) {
```

Comment:
can we remove this method? I favor immutability and this could be replaced by a constructor that takes another argument that is a map?

Reply:
will do